nil/services/synccommittee/core/fetching/aggregator.go

```diff
@@ -203,59 +211,156 @@ func (agg *aggregator) handleProcessingErr(ctx context.Context, err error) error
+		return agg.createAndPutNewBatch(ctx, &latestBatch.Id)
 	}
 
-	fetchingRange, err := types.GetBlocksFetchingRange(*startingBlockRef, *latestBlockRef, maxNumBatches)
+	agg.logger.Debug().Msgf("Latest batch with id=%s is not sealed", latestBatch.Id)
+
+	checkResult, err := agg.batchChecker.CheckConstraints(ctx, latestBatch)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("error checking batch constraints, batchId=%s: %w", latestBatch.Id, err)
 	}
 
-	if fetchingRange == nil {
-		agg.logger.Debug().
-			Stringer(logging.FieldShardId, coreTypes.MainShardId).
-			Stringer(logging.FieldBlockNumber, latestBlockRef.Number).
-			Msg("No new blocks to fetch")
-		return nil
+	switch checkResult.Type {
+	case constraints.CheckResultTypeShouldBeDiscarded:
+		agg.logger.Warn().
+			Stringer(logging.FieldBatchId, latestBatch.Id).
+			Msgf("Discarding latest batch due to constraint(s) violation: %s", checkResult.Details)
+
+		if err := agg.resetter.LaunchPartialResetWithSuspension(ctx, agg, latestBatch.Id); err != nil {
+			return nil, fmt.Errorf("error resetting progress for batch %s: %w", latestBatch.Id, err)
+		}
+
+		return nil, nil
+
+	case constraints.CheckResultTypeShouldBeSealed:
+		agg.logger.Info().
+			Stringer(logging.FieldBatchId, latestBatch.Id).
+			Msgf("Sealing batch: %s", checkResult.Details)
+
+		if err := agg.sealBatch(ctx, latestBatch); err != nil {
+			return nil, fmt.Errorf("error sealing batch: %w", err)
+		}
+
+		return nil, nil
+
+	case constraints.CheckResultTypeCanBeExtended:
+		return latestBatch, nil
+
+	default:
+		return nil, fmt.Errorf("unexpected batch check result type: %s, batchId=%s", checkResult.Type, latestBatch.Id)
 	}
+}
 
-	return agg.fetchAndProcessBlocks(ctx, *fetchingRange)
+func (agg *aggregator) createAndPutNewBatch(ctx context.Context, parentId *types.BatchId) (*types.BlockBatch, error) {
+	now := agg.clock.Now()
+	nextBatch := types.NewBlockBatch(parentId, now)
+
+	err := agg.blockStorage.PutBlockBatch(ctx, nextBatch)
+	switch {
+	case errors.Is(err, storage.ErrCapacityLimitReached):
+		return nil, fmt.Errorf("%w, cannot create new batch", err)
+
+	case errors.Is(err, context.Canceled):
+		return nil, err
+
+	case err != nil:
+		return nil, fmt.Errorf("error storing new batch, batchId=%s: %w", nextBatch.Id, err)
+
+	default:
+		return nextBatch, nil
+	}
 }
 
-// fetchLatestBlocks retrieves the latest block for main shard
-func (agg *aggregator) getLatestBlockRef(ctx context.Context) (*types.BlockRef, error) {
-	block, err := agg.rpcClient.GetBlock(ctx, coreTypes.MainShardId, "latest", false)
+func (agg *aggregator) getFetchingRange(ctx context.Context) (*types.BlocksRange, error) {
+	startingBlockRef, err := agg.getStartingBlockRef(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	latestBlockRef, err := agg.fetcher.GetLatestBlockRef(ctx, coreTypes.MainShardId)
+	if err != nil {
+		return nil, err
+	}
+
+	rangeLimit := agg.batchChecker.Constraints().MaxBlocksCount
```

Comment:
ok for now, but when we move on the block based limit, have to come back to constant fetching limit